test/evergreen/code_change_report/per_test_code_coverage_report.py

```diff
@@ -0,0 +1,228 @@
+
+            file_change_list[new_file] = change_info_list
+
+    report = dict()
+    report['change_info_list'] = file_change_list
+    report['changed_functions'] = changed_function_info
+
+    return report
+
+
+# Report on which tests reach a particular changed function
+def get_function_coverage(coverage_data: dict, changed_function: str, file_name: str, start_line: int, end_line: int,
+                          verbose: bool):
+    for test in coverage_data:
+        test_info = coverage_data[test]
+        file_coverage_data = test_info['full_coverage']['files']
+        for file_info in file_coverage_data:
+            if file_name == file_info['file']:
+                code_is_reached = False
+                for line_info in file_info['lines']:
+                    line_number = int(line_info['line_number'])
+                    line_count = int(line_info['count'])
+                    if start_line <= line_number <= end_line and line_count > 0:
+                        code_is_reached = True
+                if code_is_reached:
+                    print(f"        Reached by test: {test}")
+
+
+# Generate a report to stdout listing changed functions and which tests reach those functions
+def generate_report(coverage_data: dict, change_list: dict, preprocessed_complexity_data: dict, verbose: bool):
+    if verbose:
+        print("Generating report...")
+    report_info = create_report_info(change_list=change_list, preprocessed_complexity_data=preprocessed_complexity_data)
+    for changed_file in report_info['changed_functions']:
+        print(f'  Changed file: {changed_file}')
+        for changed_function in report_info['changed_functions'][changed_file]:
+            print(f'    Changed function: {changed_function}')
+            function_info = report_info['changed_functions'][changed_file][changed_function]
+            start_line = function_info['line start']
+            end_line = function_info['line end']
+            get_function_coverage(coverage_data=coverage_data, changed_function=changed_function,
+                                  file_name=changed_file, start_line=start_line, end_line=end_line,
+                                  verbose=verbose)
+    if verbose:
+        print("Generated report")
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-c', '--coverage_data_path', required=True, help='Path to the coverage data directory.')
+    parser.add_argument('-d', '--git_diff_file', required=True, help='Path to a Git diff file')
+    parser.add_argument('-m', '--metrix_complexity_data', required=True,
+                        help='Path to the Metrix++ complexity data csv file')
+    parser.add_argument('-v', '--verbose', action="store_true", help='Be verbose')
+    args = parser.parse_args()
+
+    verbose = args.verbose
+    coverage_data_path = args.coverage_data_path
+    git_diff_file = args.git_diff_file
+    complexity_data_file = args.metrix_complexity_data
+
+    if verbose:
```

Comment:
It looks like this would also benefit from a logger class. The idea is that we put in verbose variable, and the class would check if verbose is set, if so, we would print the message. E.g. (logger.message("hello world"), would not print if verbose is set to false.

Reply:
Good idea - I've introduced a logger.